Add ROSpan StartsWith and EndsWith string-like APIs with StringComparison

ahsonkhan opened the pull request:
Part of https://github.com/dotnet/corefx/issues/21395#issuecomment-359906138

- StartsWith
- EndsWith

Depends on: https://github.com/dotnet/coreclr/pull/16223 and https://github.com/dotnet/coreclr/pull/16418

**TODO:** ~Add tests.~

cc @jkotas, @stephentoub, @KrzysztofCwalina, @tarekgh, @JeremyKuhne

justinvp commented on src/System.Memory/src/System/MemoryExtensions.Fast.cs:
```diff
@@ -11,6 +11,24 @@ namespace System
+        /// <param name="span">The source span.</param>
+        /// <param name="value">The sequence to compare to the end of the source span.</param>
+        /// <param name="comparison">One of the enumeration values that determines how the <paramref name="span"/> and <paramref name="value"/> are compared.</param>
+        public static bool EndsWith(this ReadOnlySpan<char> span, ReadOnlySpan<char> value, StringComparison comparison)
```
`comparison` -> `comparisonType` to match the naming used everywhere else.

stephentoub commented on src/System.Memory/src/System/MemoryExtensions.Portable.cs:
```diff
@@ -12,6 +12,34 @@ namespace System
+        public static bool EndsWith(this ReadOnlySpan<char> span, ReadOnlySpan<char> value, StringComparison comparison)
+        {
+            string sourceString = span.ToString();
+            string valueString = value.ToString();
```
Ouch; this is really unfortunate.  We should consider only exposing these for core.

If we keep these, we should at least do a reasonable implementation for Ordinal and OrdinalIgnoreCase, and presumably we could do so for Invariant as well?  I agree the CurrentCulture ones will be problematic.

tarekgh replied:
Invariant would be as the current culture (if the string is not fully ASCII). 

I think there is opportunity to optimize for the ordinal and ordinal ignore casing but I think we already saying the portable implementation would be slower than the core one and whatever we do will not be able to match the core performance (i guess).

not exposing these would be an issue for anyone want to write a library that work on net core and down level too. I prefer keeping it and document it as it is slower than the core one and we can try optimize here as much we can.

---
In reply to: [166251624](https://github.com/dotnet/corefx/pull/26880#discussion_r166251624) [](ancestors = 166251624)

ahsonkhan replied:
I can file an issue to optimize this method for the ordinal cases. We do have the generic StartsWith method that take a T, which does not take SringComparison that can be used for one of the two scenarios.

FYI
>     // Those need access to globalization APIs. We'll also expose them from
    // the .NET Framework OOB (slow span). They will try to extract the string
    // from the underlying span (because slow span stores it) -- or -- allocate
    // a new string. This avoids bifurcating the API surface.

stephentoub replied:
> They will try to extract the string from the underlying span (because slow span stores it)

Can this change do this then?

ahsonkhan replied:
I haven't been able to figure out how to do that. The underlying data is stored as a Pinnable<char>, and the globalization APIs work on strings (not char*).

stephentoub replied:
It looks like we're using Unsafe.As to cast the object to a Pinnable.  Can you just do a typecheck on it and cast it back? e.g. `object o = _pinnable; if (o is string s) { ... }`?

ahsonkhan replied:
If that is doable, I can definitely update our ToString implementation and take advantage of that as well.

ahsonkhan replied:
Is there a way to do something similar for the FastSpan ToString which has a ByReference field?

jkotas replied:
No, there is not a reasonable way to do that for fast span.

jkotas commented on src/System.Memory/src/System/ReadOnlySpan.Portable.cs:
```diff
@@ -193,6 +193,10 @@ public override string ToString()
         {
             if (typeof(T) == typeof(char))
             {
+                if (_pinnable is string str)
```
This is not enough. You have to also check that the offset and length are for the whole string (ie that the Span is not pointing at substring).

jkotas commented on src/System.Memory/src/System/ReadOnlySpan.Portable.cs:
```diff
@@ -193,6 +193,10 @@ public override string ToString()
         {
             if (typeof(T) == typeof(char))
             {
+                if (_pinnable is string str)
```
This is not enough. You have to also check that the offset and length are for the whole string (ie that the Span is not pointing at substring).

ahsonkhan replied:
Would adding offset and length checks be sufficient here?

```C#
if (obj is string str && _byteOffset == 0 && _length == str.Length)
```

jkotas replied:
You need to use `StringAdjustment` for the offset.

jkotas commented on src/System.Memory/src/System/ReadOnlySpan.Portable.cs:
```diff
@@ -193,6 +193,10 @@ public override string ToString()
         {
             if (typeof(T) == typeof(char))
             {
+                if (_pinnable is string str)
```
Note that the optimizing compilers can figure out here that `Pinnable<T>` can never be a `string`, and turn this condition into false.

I think you may want to pass it through `Unsafe.As` here to strip the type information here. It is not guaranteed to work, but it is increasing the changes that it will.

jkotas replied:
We may want to have a slow-span specific test for this so that we will know when it does not work.

stephentoub replied:
> Note that the optimizing compilers can figure out here that Pinnable<T> can never be a string

Even the C# compiler will... that's why in my example I stored it into an object first, otherwise the C# compiler just emit the cast as a hard-coded false. Though as you say an even better compiler could see through the object cast as well.

ahsonkhan commented on src/System.Memory/src/System/Span.Portable.cs:
```diff
@@ -305,6 +305,10 @@ public override string ToString()
         {
             if (typeof(T) == typeof(char))
             {
+                if (_pinnable is string str)
```
Can this ever be true for Span or should I remove this check?

ahsonkhan commented:
For slow span, if we get a read/write Span from Memory, we have Pinnable\<T\> != null and hence ToString() can return the underlying string.

But, if we get construct a span from void* and length, Pinnable\<T\> == null and then ToString() returns a new string.

```C#
string original = TestHelpers.BuildString(10, 42);
ReadOnlyMemory<char> readOnlyMemory = original.AsReadOnlyMemory();
Memory<char> memory = MemoryMarshal.AsMemory(readOnlyMemory);

fixed (char* pOriginal = original)
fixed (char* pString1 = memory.Span.ToString())
{
    Assert.Equal((int)pOriginal, (int)pString1); // passes the test
}

fixed (char* pOriginal = original)
{
    Span<char> span = new Span<char>(pOriginal, original.Length);
    fixed (char* pString1 = span.ToString())
    {
        Assert.Equal((int)pOriginal, (int)pString1); // fails the test
    }
}
```

joshfree commented on src/System.Memory/tests/ReadOnlySpan/ToString.cs:
```diff
@@ -43,11 +43,53 @@ public static void ToStringChar_Empty()
+            string subString2 = span.Slice(0, 2).ToString();
+            string subString3 = span.Slice(1, 2).ToString();
+
+            Assert.Equal("RDDNEGSNET", returnedString);
```
nit: shouldn't these use `original` instead of the hardcoded literal

ahsonkhan commented:
@dotnet-bot test this please

ahsonkhan commented:
@dotnet-bot test this please

stephentoub commented on src/System.Memory/src/System/ReadOnlySpan.Portable.cs:
```diff
@@ -193,6 +193,11 @@ public override string ToString()
             if (typeof(T) == typeof(char))
             {
+                object obj = Unsafe.As<Pinnable<T>, object>(ref Unsafe.AsRef(_pinnable));
+                if (obj is string str && _byteOffset == MemoryExtensions.StringAdjustment && _length == str.Length)
```
As in https://github.com/dotnet/corefx/pull/27170/, how about doing the _byteOffset check before doing any of the casting / type checking?

ahsonkhan replied:
Will do.

stephentoub commented on src/System.Memory/tests/ReadOnlySpan/ToString.cs:
```diff
@@ -43,11 +43,56 @@ public static void ToStringChar_Empty()
+            {
+                Assert.Equal((int)pOriginal, (int)pString1);
+                Assert.Equal((int)pOriginal, (int)pString2);
+            }
```
Why not just use Assert.Same?

stephentoub commented on src/System.Memory/tests/ReadOnlySpan/ToString.cs:
```diff
@@ -43,11 +43,56 @@ public static void ToStringChar_Empty()
+                Assert.NotEqual((int)pSubString1, (int)pSubString2);
+                Assert.NotEqual((int)pSubString1, (int)pSubString3);
+                Assert.NotEqual((int)pSubString2, (int)pSubString3);
+            }
```
And Assert.NotSame?

stephentoub commented on src/System.Memory/tests/ReadOnlySpan/ToString.cs:
```diff
@@ -43,11 +43,56 @@ public static void ToStringChar_Empty()
+        // This test is only relevant for portable span
+#if FEATURE_PORTABLE_SPAN
+        [Fact]
+#endif
```
Is this the convention used in this project?  Elsewhere in corefx we've used SkipOnTargetFramework, e.g. https://github.com/dotnet/corefx/pull/27170/files#diff-f8d6f156088e92893813570d9608526eR63

ahsonkhan replied:
I am not sure if NetFramework is the only one TargetFrameworkMoniker where we would skip this test.

stephentoub replied:
Ok.

ahsonkhan replied:
But given [this](https://github.com/dotnet/corefx/blob/7059f8b316723f5db64e4b1d28e2b1e8b137ab1e/Documentation/project-docs/developer-guide.md#skipontargetframeworkattribute), I will change it:

stephentoub commented on src/System.Memory/src/System/ReadOnlySpan.Portable.cs:
```diff
@@ -193,6 +193,11 @@ public override string ToString()
         {
             if (typeof(T) == typeof(char))
             {
+                object obj = Unsafe.As<Pinnable<T>, object>(ref Unsafe.AsRef(_pinnable));
```
Can this just be:
```C#
object obj = Unsafe.As<object>(_pinnable);
```
?

stephentoub replied:
Nit: (Also, a comment indicating why this is used rather than just `if (_pinnable is string str ...)` would be helpful.)

ahsonkhan replied:
I will merge the changes from your PR into here which will include the comments, test updates, etc.

Thanks!

stephentoub replied:
Cool, thanks.  Didn't realize you'd already done the initial work (I remember we'd talked about it, but then couldn't find it in the repo :smile:)

stephentoub commented on src/System.Memory/src/System/Span.Portable.cs:
```diff
@@ -305,6 +305,11 @@ public override string ToString()
+                if (obj is string str && _byteOffset == MemoryExtensions.StringAdjustment && _length == str.Length)
+                {
+                    return str;
+                }
```
Is this change needed?  When will a `Span<char>` (rather than a `ReadOnlySpan<char>`) wrap a `string`?

ahsonkhan replied:
See code sample here: https://github.com/dotnet/corefx/pull/26880#issuecomment-363632190

stephentoub replied:
I personally don't think it's worthwhile to add this check to `Span<char>`.  It's such a corner case to optimize for, and at the expense of adding a type check to other `Span<char>.ToString()` calls.  And in general folks can't depend on ToString producing the same string from multiple calls, so I don't consider the discrepancy a real issue.

ahsonkhan replied:
I don't have a strong opinion on this. I am fine with removing this check for Span. @KrzysztofCwalina, @jkotas - any objections?

ahsonkhan commented:
@dotnet-bot test Linux x64 Release Build

ahsonkhan commented:
@dotnet-bot test this please
